server/src/main/java/io/druid/segment/realtime/firehose/ReplayableFirehoseFactory.java

```diff
@@ -0,0 +1,313 @@
+      Set<String> dimensionScratch = new HashSet<>();
+
+      File tmpDir = Files.createTempDir();
+      tmpDir.deleteOnExit();
+
+      long startTime = System.nanoTime();
+      do {
+        deleteTempFiles();
+        try (Firehose delegateFirehose = delegateFactory.connect(parser)) {
+          while (delegateFirehose.hasMore()) {
+            File tmpFile = File.createTempFile("replayable-", null, tmpDir);
+            tmpFile.deleteOnExit();
+
+            files.add(tmpFile);
+            log.debug("Created file [%s]", tmpFile.getAbsolutePath());
+
+            try (CountingOutputStream cos = new CountingOutputStream(new FileOutputStream(tmpFile));
+                 JsonGenerator generator = factory.createGenerator(cos)) {
+
+              while (delegateFirehose.hasMore() && cos.getCount() < getMaxTempFileSize()) {
+                try {
+                  InputRow row = delegateFirehose.nextRow();
+                  generator.writeObject(row);
+                  dimensionScratch.addAll(row.getDimensions());
+                  counter++;
+                }
+                catch (ParseException e) {
+                  if (reportParseExceptions) {
+                    throw e;
+                  }
+                  unparseable++;
+                }
+              }
+
+              totalBytes += cos.getCount();
+            }
+          }
+          break; // we succeeded so break out of retry loop
+        }
+        catch (Exception e) {
+          if (++retryCount <= readFirehoseRetries && !(e instanceof ParseException)) {
+            log.error(e, "Delegate firehose threw an exception, retrying (%d of %d)", retryCount, readFirehoseRetries);
+          } else {
+            log.error(e, "Delegate firehose threw an exception, retries exhausted, aborting");
+            Throwables.propagate(e);
+          }
+        }
+      } while (true);
+
+      log.info(
+          "Finished reading from firehose in [%,dms], [%,d] events parsed, [%,d] bytes written, [%,d] events unparseable",
+          (System.nanoTime() - startTime) / 1000000,
+          counter,
+          totalBytes,
+          unparseable
+      );
+
+      dimensions = Ordering.natural().immutableSortedCopy(dimensionScratch);
+
+      if (counter == 0) {
+        log.warn("Firehose contains no events!");
+        deleteTempFiles();
+        it = Iterators.emptyIterator();
+      } else {
+        jsonParser = factory.createParser(files.get(fileIndex));
+        it = jsonParser.readValuesAs(Row.class);
+      }
+    }
+
+    @Override
+    public boolean hasMore()
+    {
+      return it.hasNext() || (fileIndex < files.size() - 1);
+    }
+
+    @Override
+    public InputRow nextRow()
+    {
+      if (!it.hasNext()) {
+        try {
+          CloseQuietly.close(jsonParser);
+
+          if (++fileIndex >= files.size()) {
+            throw new IllegalStateException("No more data!");
+          }
+
+          jsonParser = factory.createParser(files.get(fileIndex));
```

Comment:
will things be ok, if the file returned from files.get(fileIndex) is empty. it is possible (a corner case though) for the last file to not have any data in it.